src/main/java/org/vanilladb/core/storage/buffer/Buffer.java

```diff
@@ -295,16 +289,14 @@ boolean isModified() {
 	 *            a block ID
 	 */
 	void assignToBlock(BlockId blk) {
-		internalLock.writeLock().lock();
-		try {
-			flush();
-			this.blk = blk;
-			contents.read(blk);
-			pins = 0;
-			lastLsn = LogSeqNum.readFromPage(contents, LAST_LSN_OFFSET);
-		} finally {
-			internalLock.writeLock().unlock();
-		}
+		// Optimization: This might be a danger optimization
+		// This method is called because no tx pin this buffer,
+		// which means no tx will modify or read the content.
+		flush();
+		this.blk = blk;
+		contents.read(blk);
+		pins = new AtomicInteger(0);
+		lastLsn = LogSeqNum.readFromPage(contents, LAST_LSN_OFFSET);

```

Comment:
Maybe we can add a check at the beginning to ensure there is no one pinning this buffer.

Reply:
I would like to throw a RunTimeException if pin > 0.
Let it crash first, then fix it.

Reply:
Sure